rust/operator-binary/src/controller.rs

```diff
@@ -0,0 +1,1002 @@
+                .take(15)
+                .map(char::from)
+                .collect();
+
+            let mut secret_data = BTreeMap::new();
+            secret_data.insert("nifiSensitivePropsKey".to_string(), password);
+
+            let new_secret = Secret {
+                metadata: ObjectMetaBuilder::new()
+                    .namespace(namespace)
+                    .name(&sensitive_config.key_secret.to_string())
+                    .build(),
+                string_data: Some(secret_data),
+                ..Secret::default()
+            };
+            client
+                .create(&new_secret)
+                .await
+                .with_context(|_| SensitiveKeySecretSnafu {})?;
+            Ok(true)
+        }
+    }
+}
+
+fn get_stackable_secret_volume_attributes() -> BTreeMap<String, String> {
+    let mut result = BTreeMap::new();
+    result.insert(
+        "secrets.stackable.tech/class".to_string(),
+        "tls".to_string(),
+    );
+    result.insert(
+        "secrets.stackable.tech/scope".to_string(),
+        "node,pod".to_string(),
+    );
+    result
+}
+
+fn build_persistent_volume_claim_rwo_storage(name: &str, storage: &str) -> PersistentVolumeClaim {
+    PersistentVolumeClaim {
+        metadata: ObjectMeta {
+            name: Some(name.to_string()),
+            ..ObjectMeta::default()
+        },
+        spec: Some(PersistentVolumeClaimSpec {
+            access_modes: Some(vec!["ReadWriteOnce".to_string()]),
+            resources: Some(ResourceRequirements {
+                requests: Some({
+                    let mut map = BTreeMap::new();
+                    map.insert("storage".to_string(), Quantity(storage.to_string()));
+                    map
+                }),
+                ..ResourceRequirements::default()
+            }),
+            ..PersistentVolumeClaimSpec::default()
+        }),
+        ..PersistentVolumeClaim::default()
+    }
+}
+
+async fn external_node_port(nifi_service: &Service) -> Result<i32> {
+    let external_ports = nifi_service
+        .spec
+        .as_ref()
+        .with_context(|| ObjectHasNoSpecSnafu {})?
+        .ports
+        .as_ref()
+        .with_context(|| ExternalPortSnafu {})?
+        .iter()
+        .filter(|p| p.name == Some(HTTPS_PORT_NAME.to_string()))
+        .collect::<Vec<_>>();
+
+    let port = external_ports
+        .first()
+        .with_context(|| ExternalPortSnafu {})?;
+
+    Ok(port.node_port.with_context(|| ExternalPortSnafu {})?)
+}
+
+async fn get_proxy_hosts(
+    client: &Client,
+    nifi: &NifiCluster,
+    nifi_service: &Service,
+) -> Result<String> {
+    let selector = LabelSelector {
+        match_labels: {
+            let mut labels = BTreeMap::new();
+            labels.insert("kubernetes.io/os".to_string(), "linux".to_string());
+            Some(labels)
+        },
+        ..LabelSelector::default()
+    };
+
+    let external_port = external_node_port(nifi_service).await?;
```

Comment:
`external_node_port()` doesn't do any i/o. It need not be `async`.

Reply:
fixed